### PR TITLE
Adapt Http2ServerDowngrader to work with clients

### PR DESCRIPTION
**Pull request by lionelli1**
Motivation:
Http2ServerDowngrader is specifically built for server channels where
inbound Http2StreamFrames are converted into HttpRequests, and outbound
HttpResponses are converted into Http2StreamFrames. It can be easily
made to be more generic to work with client channels where inbound
Http2StreamFrames are converted into HttpResponses, and outbound
HttpRequests are converted into Http2StreamFrames.

Modification:
- Renamed Http2ServerDowngrader to a more general
  Http2StreamFrameToHttpObjectCodec
- Made it take in an "isServer" parameter to determine whether encoding
  inbound Http2StreamFrames should create HttpRequests (for server) or
  HttpResponses (for client)

Result:
Now Http2StreamFrameToHttpObjectCodec can be used to translate
Http2StreamFrame to HttpObject for both server and client.

Fixes https://github.com/netty/netty/issues/7171

### CODE REVIEW

**Comment by lionelli1** [comment]
@normanmaurer @Scottmitch PTAL thanks!

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java [hunk 1]
```diff
@@ -16,46 +16,58 @@
...
-public class Http2ServerDowngrader extends MessageToMessageCodec<Http2StreamFrame, HttpObject> {
-
+public class Http2StreamFrameToHttpObjectCodec extends MessageToMessageCodec<Http2StreamFrame, HttpObject> {
+    private final boolean isServer;
     private final boolean validateHeaders;
 
-    public Http2ServerDowngrader(boolean validateHeaders) {
+    private volatile HttpScheme scheme;
+
+    public Http2StreamFrameToHttpObjectCodec(final boolean isServer,
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: @lionelli1 nit: either make both final or none of these :)

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java [hunk 2]
```diff
@@ -71,18 +83,17 @@ protected void decode(ChannelHandlerContext ctx, Http2StreamFrame frame, List<Ob
             Http2Headers headers = headersFrame.headers();
 
             if (headersFrame.isEndStream()) {
-                if (headers.method() == null) {
+                if ((headers.method() == null) && (headers.status() == null)) {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: remove `(...)` around checks ?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java [hunk 3]
```diff
@@ -132,4 +143,44 @@ protected void encode(ChannelHandlerContext ctx, HttpObject obj, List<Object> ou
...
+
+        if (isSsl(ctx.channel())) {
+            scheme = HttpScheme.HTTPS;
+        }
+    }
+
+    protected boolean isSsl(final Channel ch) {
+        final Channel channel = (ch instanceof Http2StreamChannel) ? ch.parent() : ch;
+        final ChannelPipeline p = channel.pipeline();
+        return (null != p.get(SslHandler.class)) || (null != p.get(SniHandler.class));
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: remove `(...)` around the checks its not needed.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java [hunk 4]
```diff
@@ -132,4 +143,44 @@ protected void encode(ChannelHandlerContext ctx, HttpObject obj, List<Object> ou
...
+        return isServer ?
+                HttpConversionUtil.toFullHttpRequest(id, headers, alloc, validateHeaders) :
+                HttpConversionUtil.toFullHttpResponse(id, headers, alloc, validateHeaders);
+    }
+
+    @Override
+    public void handlerAdded(final ChannelHandlerContext ctx) throws Exception {
+        super.handlerAdded(ctx);
+
+        if (isSsl(ctx.channel())) {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: maybe just use:

```java
scheme = isSsl(...) ? HttpScheme.HTTPS : HttpScheme.HTTP;
```

This way its more explicit about what is used and what not

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java [hunk 5]
```diff
@@ -132,4 +143,44 @@ protected void encode(ChannelHandlerContext ctx, HttpObject obj, List<Object> ou
...
+            scheme = HttpScheme.HTTPS;
+        }
+    }
+
+    protected boolean isSsl(final Channel ch) {
```
Comment (normanmaurer) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: make this take `ChannelHandlerContext`

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java (hunk 1 above)
Comment (lionelli1) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: oops, missed that one out. done!

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java (hunk 2 above)
Comment (lionelli1) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: removed

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java (hunk 4 above)
Comment (lionelli1) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: yup, changed

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java (hunk 5 above)
Comment (lionelli1) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: done

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java (hunk 3 above)
Comment (lionelli1) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: removed

**Comment by lionelli1** [comment]
@normanmaurer i've addressed your review comments. back to you. thanks!

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java [hunk 6]
```diff
@@ -16,46 +16,58 @@
...
  */
 @UnstableApi
-public class Http2ServerDowngrader extends MessageToMessageCodec<Http2StreamFrame, HttpObject> {
-
+public class Http2StreamFrameToHttpObjectCodec extends MessageToMessageCodec<Http2StreamFrame, HttpObject> {
+    private final boolean isServer;
     private final boolean validateHeaders;
 
-    public Http2ServerDowngrader(boolean validateHeaders) {
+    private volatile HttpScheme scheme;
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: it looks like this variable is always access in the EventLoop thread ... does this have to be volatile?

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java [hunk 7]
```diff
@@ -132,4 +143,43 @@ protected void encode(ChannelHandlerContext ctx, HttpObject obj, List<Object> ou
...
+        super.handlerAdded(ctx);
+
+        scheme = isSsl(ctx) ? HttpScheme.HTTPS : HttpScheme.HTTP;
+    }
+
+    protected boolean isSsl(final ChannelHandlerContext ctx) {
+        final Channel ch = ctx.channel();
+        final Channel connChannel = (ch instanceof Http2StreamChannel) ? ch.parent() : ch;
+        final ChannelPipeline p = connChannel.pipeline();
+        return null != p.get(SslHandler.class) || null != p.get(SniHandler.class);
```
Comment (Scottmitch) [inline codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: if SNI is in the pipeline we should wait until a `SslHandshakeCompletionEvent` is received in the pipeline before we consider the connection to be `HTTPS`.

For example:
```java
@Override
public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
  if (evt instanceof SslHandshakeCompletionEvent) {
    // SslHandler should be in the pipeline
  }
}
```

File: codec-http2/src/test/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodecTest.java [hunk 8]
```diff
@@ -0,0 +1,799 @@
...
+            goaway.release();
+            frame.release();
+        }
+    }
+
+    private SslHandler buildSsl() throws Exception {
+        final SslContext ctx = SslContextBuilder.forClient().build();
+
+        // this SslHandler bypasses Http2StreamFrames so we can verify the
+        // object written
```
Comment (Scottmitch) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodecTest.java]: nit: put this line on the previous line

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java (hunk 6 above)
Comment (normanmaurer) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: +1... not need for volatile

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java (hunk 7 above)
Comment (normanmaurer) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: +1

File: codec-http2/src/test/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodecTest.java (hunk 8 above)
Comment (normanmaurer) [reply codec-http2/src/test/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodecTest.java]: +1

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java (hunk 7 above)
Comment (normanmaurer) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: @lionelli1 @Scottmitch that said we need to ensure we do this as part of the "correct" pipeline which may be the parent pipeline or not depending on if `Http2MultiplexCodex` is used or not.

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java (hunk 6 above)
Comment (lionelli1) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: good point. removed volatile

File: codec-http2/src/test/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodecTest.java (hunk 8 above)
Comment (lionelli1) [reply codec-http2/src/test/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodecTest.java]: done

File: codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java (hunk 7 above)
Comment (lionelli1) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: but for most of the use cases where this handler is added at the child stream, the SslHandshakeCompletionEvent would've happened at the parent channel by the time it is added to the stream channels. the new child channels won't be notified of this event upon creation, would it?

Comment (normanmaurer) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: yep ... maybe its ok for now to just do this only if the its not a "child channel" for now

Comment (lionelli1) [reply codec-http2/src/main/java/io/netty/handler/codec/http2/Http2StreamFrameToHttpObjectCodec.java]: ok changed to checking on SslHandler with the assumption that this handler will mostly be used on an `Http2StreamChannel`

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (e069079aff5111b45d333bcce5432260fad99076)

**Pull request closed without merging**